### PR TITLE
feat: add `ARRAY_CONCAT` UDF

### PR DESCRIPTION
**Pull request by patrickstuedi**
### Description 

Adding `array_concat` UDF function. 

DESCRIBE function array_concat:
```
Name        : ARRAY_CONCAT
Overview    : Concatenates two arrays, creating an array that contains all the elementsin the first array followed by all the elements in the second array.
Type        : SCALAR
Jar         : internal
Variations  : 

        Variation   : ARRAY_CONCAT(left ARRAY<T>, right ARRAY<T>)
        Returns     : ARRAY<T>
        left        : First array of values
        right       : Second array of values
```

### Testing done 

Tested with different array values, including duplicates, null value array members and null value input parameters

### CODE REVIEW

**Review by JimGalasyn — APPROVED** [review]
LGTM!

File: ksqldb-engine/src/test/java/io/confluent/ksql/function/udf/array/ArrayConcatTest.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+public class ArrayConcatTest {
+
+  private final ArrayConcat udf = new ArrayConcat();
+
+  @Test
+  public void shouldConcatArraysOfLikeType() {
+    final List<String> input1 = Arrays.asList("foo", " ", "bar");
+    final List<String> input2 = Arrays.asList("baz");
+    final List<String> result = udf.concat(input1, input2);
+    assertThat(result.size(), is(input1.size()+input2.size()));
```
Comment (vvcephei) [inline ksqldb-engine/src/test/java/io/confluent/ksql/function/udf/array/ArrayConcatTest.java:37]:
```suggestion
    assertThat(result, is(Arrays.asList("foo", " ", "bar", "baz"));
```

File: ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/array/ArrayConcat.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        + "The two arrays must be of the same type.")
+public class ArrayConcat {
+  @Udf
+  public <T> List<T> concat(
+      @UdfParameter(description = "First array of values") final List<T> left,
+      @UdfParameter(description = "Second array of values") final List<T> right) {
+    if (left == null && right == null) {
+      return null;
+    }
+    final List<T> result = Lists.newArrayList();
```
Comment (vvcephei) [inline ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/array/ArrayConcat.java:40]: Let's initialize this arrayList to be size "left.size + right.size": `new ArrayList(left.size() + right.size())`

**Review by vvcephei — COMMENTED** [review]
Thanks for this! Just a couple of thoughts.

File: ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/array/ArrayConcat.java [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+        + "The two arrays must be of the same type.")
+public class ArrayConcat {
+  @Udf
+  public <T> List<T> concat(
+      @UdfParameter(description = "First array of values") final List<T> left,
+      @UdfParameter(description = "Second array of values") final List<T> right) {
+    if (left == null && right == null) {
+      return null;
+    }
+    final List<T> result = new ArrayList(left.size() + right.size());
```
Comment (agavra) [inline ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/array/ArrayConcat.java:40]: if `left == null || right == null` wouldn't we get an NPE here?

File: ksqldb-functional-tests/src/test/resources/query-validation-tests/array-set-functions.json [hunk 4]
```diff
@@ -171,6 +171,44 @@
...
+      ]
+    },
+    {
+      "name": "array_concat with primitive type",
+      "statements": [
+        "CREATE STREAM INPUT (id STRING KEY, arr1 ARRAY<INT>, arr2 ARRAY<INT>) WITH (kafka_topic='test_topic', value_format='JSON');",
+        "CREATE STREAM OUTPUT AS SELECT id, array_concat(arr1, arr2) as result FROM INPUT;"
+      ],
+      "inputs": [
+        {"topic": "test_topic", "key": "r1", "value": {"arr1": [0,0,1,0,-1], "arr2": [1,-2,0]}},
```
Comment (agavra) [inline ksqldb-functional-tests/src/test/resources/query-validation-tests/array-set-functions.json:197]: let's also add the null cases here (left is null, right is null, both are null)

**Review by agavra — COMMENTED** [review]
took a quick scan of the PR - thanks @patrickstuedi!

File: ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/array/ArrayConcat.java (hunk 3 above)
Comment (vvcephei) [reply ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/array/ArrayConcat.java:40]: I thought so too, but there are unit tests for those cases. Not sure what to make of that.

File: docs/developer-guide/ksqldb-reference/scalar-functions.md [hunk 5]
```diff
@@ -426,6 +426,24 @@ ARRAY_UNION(ARRAY[1, 2, 3, 1, 2], [4, 1])  => [1, 2, 3, 4]
...
+
+Since: 0.20.0
+
+```sql
+ARRAY_CONCAT(array1, array2)
+```
+
+Returns an array representing the concatenation of both input arrays.
+
+Returns NULL if both input arrays are NULL.
```
Comment (vvcephei) [inline docs/developer-guide/ksqldb-reference/scalar-functions.md:439]:
```suggestion
Returns NULL if both input arrays are NULL. If only one argument is NULL, the result is the other argument.
```

... Or any other way you prefer to say it, but we should specify the behavior if just one arg is null.

File: ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/array/ArrayConcat.java (hunk 3 above)
Comment (patrickstuedi) [reply ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/array/ArrayConcat.java:40]: thanks for pointing out. missed that when changing the initialization to the exact number of elements. No idea what's going on with the tests, they should fail here. Will double check.

Comment (patrickstuedi) [reply ksqldb-engine/src/main/java/io/confluent/ksql/function/udf/array/ArrayConcat.java]: Tests did catch this, sorry for missing to run them on that last change. Update fixes the issue in ArrayConcat.

**Comment by vvcephei** [comment]
I just checked on the test failures: https://jenkins.confluent.io/job/Confluentinc%20Contributors/job/ksql/job/PR-7761/7/#showFailuresLink

It looks like the same set of tests started failing on master yesterday: https://jenkins.confluent.io/job/confluentinc/job/ksql/job/master/7992/#showFailuresLink

**Review by vvcephei — APPROVED** [review]
Thanks @patrickstuedi !

**Pull request merged**